abracadabra/src/writer/persistency_manager.cc

```diff
@@ -0,0 +1,64 @@
+#include "persistency_manager.hh"
+#include "g4-mandatory/trajectory.hh"
+
+
+persistency_manager::persistency_manager() {
+    G4String hdf5file = "test_file.h5";
+    nevt = 0;
+    open_file(hdf5file);
+}
```

Comment:
> Therefore, by hard-wiring opening of the output file in this constructor, we are limiting ourselves to just one write per process. (The usual Geant4-style inflexibility.) We should avoid this restriction if at all possible.

We can move the opening of the file to some other place (where?), but I don't see any scenario in which we will need opening more than one output file in the same job.

> (As you mentioned elsewhere, this hard-wired name has to be made more flexible.)

I'm aware of the hard-coded name issue. At some point it will be read from the configuration file using a G4 messenger, but I do not know yet how to do that.

Reply:
> I don't see any scenario in which we will need opening more than one output file in the same job.
 
For a start, we will want to open many outputs in the single process which runs our extensive suite of tests, and gathers data about them.
 
Making code easily testable tends to lead to loosely coupled, orthogonal, easily composable and reusable and easily understandable and debuggable code. If your code isn't easily testable, then it's probably badly designed.
 
The immense difficulty of testing Geant4 code is a strong indicator of how badly designed it is.
 
> using a G4 messenger

Aaah, yes, the G4 messenger. Beware. Pretty much all G4 messenger examples you'll find in the wild, are a horrid exercise in mutating shared state (because G4 itself is one huge exercise in mutating shared state). We must avoid falling into the trap of repeating the awful usage patterns that are standard practice around the G4 messenger.

We need to establish some tolerable practices in this area. I haven't yet had the time to give this careful thought, but it is likely to be based on mutating some configuration object (from what I've seen so far, mutation is the *only* way of doing anything with the messenger: I hope I'm wrong, but I don't hold out much hope) and using the data contained in that object to construct immutable objects that will allow us to preserve our sanity.